testgen/src/tester.rs

```diff
@@ -1,77 +1,196 @@
 use crate::helpers::*;
 use crate::tester::TestResult::{Failure, ParseError, ReadError, Success};
 use serde::de::DeserializeOwned;
-use std::panic::UnwindSafe;
-use std::{fs, path::PathBuf};
 use std::{
-    panic,
+    fs,
+    io::Write,
+    panic::{self, UnwindSafe},
+    path::{Path, PathBuf},
     sync::{Arc, Mutex},
 };
+use tempfile::TempDir;
 
+/// A test environment, which is essentially a wrapper around some directory,
+/// with some utility functions operating relative to that directory.
 #[derive(Debug, Clone)]
 pub struct TestEnv {
-    root_dir: String,
-    logs: Vec<String>,
+    /// Directory where the test is being executed
+    current_dir: String,
 }
 
 impl TestEnv {
-    pub fn add_log(&mut self, log: &str) {
-        self.logs.push(log.to_string());
+    pub fn new(current_dir: &str) -> Option<Self> {
+        fs::create_dir_all(current_dir).ok().map(|_| TestEnv {
```

Comment:
I wonder if we might need a sanity check here to prevent users passing `/`, `~`, or `.`, and then end up with their eg. home directory erased in the `cleanup` step?

Reply:
That's true; I will add the checks

Reply:
Perhaps just appending a directory to the given path is sufficient and even more reliable. This way, whatever path we are given, we will always store temp files one directory below and will thus not touch/override/remove whatever content there might already be within the given directory.

Reply:
the last solution will not work, as I need this also to refer to and copy existing files from the test "root" directory. I think the better solution is not to check at all, and move the `cleanup` function away from the `TestEnv`, so no destructive actions will be available to the user. Instead I will do cleanup exclusively in the `Tester` itself, and only for the temporary dir it creates

Reply:
Sounds like the best solution altogether!